federation/pkg/federation-controller/service/service_helper.go

```diff
@@ -0,0 +1,219 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"time"
+	"fmt"
+
+	"k8s.io/kubernetes/federation/client/clientset_generated/federation_release_1_3"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/v1"
+	cache "k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/golang/glog"
+	"reflect"
+)
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (sc *ServiceController) clusterServiceWorker() {
+	fedClient := sc.federationClient
+	for clusterName, clusterCache := range sc.clusterCache.clientMap{
+		go func() {
+			for {
+				key, quit := clusterCache.serviceQueue.Get()
+				// update service cache
+				if quit {
+					return
+				}
+				defer clusterCache.serviceQueue.Done(key)
+				err := sc.clusterCache.syncService(key.(string), clusterName, clusterCache, sc.serviceCache, fedClient)
+				if err != nil {
+					glog.V(2).Infof("failed to sync service: %+v", err)
+					//return err
+				}
+			}
+		}()
+	}
+}
+
+// Whenever there is change on service, the federation service should be updated
+func (cc *clusterClientCache) syncService(key, clusterName string, clusterCache *clusterCache, serviceCache *serviceCache, fedClient federation_release_1_3.Interface) error {
+	// obj holds the latest service info from apiserver, return if there is no federation cache for the service
+	cachedService, ok := serviceCache.get(key)
+	if !ok {
+		// here we filtered all non-federation services
+		return nil
+	}
+	obj, exists, err := clusterCache.serviceStore.GetByKey(key)
+	if err != nil {
+		glog.Infof("Unable to retrieve service %v from store: %v", key, err)
+		clusterCache.serviceQueue.Add(key)
+		return err
+	}
+	var needUpdate bool
+	if exists {
+		service, ok := obj.(*v1.Service)
+		if ok {
+			glog.V(4).Infof("Found service for federation service %s/%s from cluster %s", service.Namespace, service.Name, clusterName)
+			needUpdate = cc.processServiceUpdate(cachedService, service, clusterName)
+		} else {
+			tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+			if !ok {
+				return fmt.Errorf("object contained wasn't a service or a deleted key: %+v", obj)
+			}
+			glog.Infof("Found tombstone for %v", key)
+			needUpdate = cc.processServiceDeletion(cachedService, tombstone.Key, clusterName)
+		}
+	} else {
+		// service absence in store means watcher caught the deletion
+		glog.Infof("service has been deleted %v", key)
+		needUpdate = cc.processServiceDeletion(cachedService, key, clusterName)
+	}
+	if needUpdate {
+		cc.persistFedServiceUpdate(cachedService.lastState, fedClient)
+		cachedService.appliedState = cachedService.lastState
```

Comment:
My understanding is that DNS update occurs based on endpoint worker, while k8s service worker only update the federation service info to federation apiserver. so here we should not update dns records.